src/carma-1-tenth/plot_localization.py

```diff
@@ -0,0 +1,169 @@
+# Plot the standard deviation of the vehicle's localization as a function of downtrack distance
+
+
+from rosbag_utils import open_bagfile
+from plot_crosstrack_error import find_closest_point
+import numpy as np
+import yaml
+import tqdm
+from rosidl_runtime_py.utilities import get_message
+from rclpy.serialization import deserialize_message
+import matplotlib.pyplot as plt
+import datetime
+from scipy.interpolate import make_interp_spline, interp1d
+import argparse, argcomplete
+import os
+
+
+def plot_localization(bag_dir, start_offset=0.0):
+    metadatafile : str = os.path.join(bag_dir, "metadata.yaml")
+    if not os.path.isfile(metadatafile):
+        raise ValueError("Metadata file %s does not exist. Are you sure %s is a valid rosbag?" % (metadatafile, bag_dir))
+    with open(metadatafile, "r") as f:
+        metadata_dict : dict = yaml.load(f, Loader=yaml.SafeLoader)["rosbag2_bagfile_information"]
+    storage_id = metadata_dict['storage_identifier']
+    odom_topic = '/amcl_pose'
+    route_topic = '/route_graph'
+    particles_topic = '/particle_cloud'
+    cmd_vel_topic = '/cmd_vel'
+    reader, type_map = open_bagfile(bag_dir, topics=[odom_topic, route_topic, particles_topic, cmd_vel_topic], storage_id=storage_id)
+    topic_count_dict = {entry["topic_metadata"]["name"] : entry["message_count"] for entry in metadata_dict["topics_with_message_count"]}
+
+    route_graph = None
+    odom_count = 0
+    particles_count = 0
+    velocities_count = 0
+    odometry = np.zeros((topic_count_dict[odom_topic], 2))
+    odometry_times = np.zeros((topic_count_dict[odom_topic],))
+    particle_std_deviations = np.zeros((topic_count_dict[particles_topic],))
+    particle_times = np.zeros((topic_count_dict[particles_topic],))
+    velocities = np.zeros((topic_count_dict[cmd_vel_topic],))
+    velocity_times = np.zeros((topic_count_dict[cmd_vel_topic],))
+    # Iterate through bag and store odometry + route_graph messages
+    for idx in tqdm.tqdm(iterable=range(topic_count_dict[odom_topic] + topic_count_dict[route_topic] + topic_count_dict[particles_topic])):
+        if(reader.has_next()):
+            (topic, data, t_) = reader.read_next()
+            msg_type = type_map[topic]
+            msg_type_full = get_message(msg_type)
+            msg = deserialize_message(data, msg_type_full)
+            if topic == route_topic:
+                route_graph = msg
+            elif topic == odom_topic:
+                odometry[odom_count] = [-msg.pose.pose.position.y, msg.pose.pose.position.x]
```

Comment:
Could you add a comment on the 90 degree rotation that this implies

Reply:
This is done to display the vehicle odometry readings + route in the x-forward, y-left coordinate convention instead of standard x,y coordinates. I just added a comment to indicate this